### PR TITLE
`Development`: Make small improvements for build job queue in local continuous integration

### PR DESCRIPTION
**Pull request by laurenzfb**
…ildJobQueueItem into records

#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [ ] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [ ] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [ ] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [ ] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [ ] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [ ] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [ ] I documented the Java code using JavaDoc style.
#### Client
- [ ] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [ ] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [ ] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [ ] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [ ] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [ ] I documented the TypeScript code using JSDoc style.
- [ ] I added multiple screenshots/screencasts of my UI changes.
- [ ] I translated all newly inserted strings into English and German.
#### Changes affecting Programming Exercises
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Programming Exercise with Complaints enabled

1. Log in to Artemis
2. Navigate to Course Administration
3. ...

#### Exam Mode Testing

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Exam with a Programming Exercise

1. Log in to Artemis
2. Participate in the exam as a student
3. Make sure that the UI of the programming exercise in the exam mode stays unchanged. You can use the [exam mode documentation](https://docs.artemis.cit.tum.de/user/exam_mode/) as reference.
4. ...

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/config/localvcci/LocalCIConfiguration.java [hunk 1]
```diff
@@ -42,10 +42,10 @@ public class LocalCIConfiguration {
     @Value("${artemis.continuous-integration.docker-connection-uri}")
     String dockerConnectionUri;
 
-    @Value("${artemis.continuous-integration.thread-pool-size:1}")
+    @Value("${artemis.continuous-integration.concurrent-build-size:1}")
     int fixedThreadPoolSize;
```
Comment (reschandreas) [inline src/main/java/de/tum/in/www1/artemis/config/localvcci/LocalCIConfiguration.java:46]: maybe also rename the variables to stay consistent

File: src/main/java/de/tum/in/www1/artemis/config/localvcci/LocalCIConfiguration.java [hunk 2]
```diff
@@ -42,10 +42,10 @@ public class LocalCIConfiguration {
...
 
-    @Value("${artemis.continuous-integration.specify-thread-pool-size:false}")
+    @Value("${artemis.continuous-integration.specify-concurrent-builds:false}")
     boolean specifyThreadPoolSize;
```
Comment (reschandreas) [inline src/main/java/de/tum/in/www1/artemis/config/localvcci/LocalCIConfiguration.java:49]: same here

File: src/main/java/de/tum/in/www1/artemis/config/websocket/WebsocketConfiguration.java [hunk 3]
```diff
@@ -268,6 +270,32 @@ public Message<?> preSend(@NotNull Message<?> message, @NotNull MessageChannel c
...
+            }
+            // Define a pattern to match the expected course-related topic format
+            Pattern pattern = Pattern.compile("^/topic/courses/(\\d+)/(queued-jobs|running-jobs)$");
+            Matcher matcher = pattern.matcher(destination);
+
+            // Check if the destination matches the pattern
+            if (matcher.matches()) {
+                // Extract the courseId from the matched groups
+                String courseIdString = matcher.group(1);
+                long courseId = Long.parseLong(courseIdString);
```
Comment (reschandreas) [inline src/main/java/de/tum/in/www1/artemis/config/websocket/WebsocketConfiguration.java:288]:
```suggestion
                long courseId = Long.parseLong( matcher.group(1));
```

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCISharedBuildJobQueueService.java [hunk 4]
```diff
@@ -99,6 +103,10 @@ public LocalCISharedBuildJobQueueService(HazelcastInstance hazelcastInstance, Ex
         this.processingJobs.addLocalEntryListener(new ProcessingBuildJobItemListener());
         this.buildAgentInformation.addLocalEntryListener(new BuildAgentListener());
         this.localCIBuildQueueWebsocketService = localCIBuildQueueWebsocketService;
+
+        if (env.acceptsProfiles(Profiles.of("scheduling"))) {
```
Comment (reschandreas) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCISharedBuildJobQueueService.java:107]: you could add a helper method in the ProfileService like the `isBamboo` or `isLocalVcsCi` and shorten this to `this.schedulingActive = profileService.isScheduling()`. But that is nonblocking for me

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCISharedBuildJobQueueService.java [hunk 5]
```diff
@@ -340,11 +351,11 @@ private void processBuild(LocalCIBuildJobQueueItem buildJob) {
         }).exceptionally(ex -> {
             log.error("Error while processing build job: {}", buildJob, ex);
 
-            processingJobs.remove(buildJob.getId());
+            processingJobs.remove(buildJob.id());
             localProcessingJobs.decrementAndGet();
             updateLocalBuildAgentInformation();
 
-            if (buildJob.getRetryCount() > 0) {
+            if (buildJob.retryCount() > 0) {
```
Comment (reschandreas) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCISharedBuildJobQueueService.java:358]: could you use `isRetry` here? not sure but would make sense imho

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCISharedBuildJobQueueService.java [hunk 6]
```diff
@@ -406,57 +419,71 @@ private ProgrammingExerciseParticipation retrieveParticipationWithRetry(Long par
...
+                log.debug("Build agent added: {}", event.getValue());
+                localCIBuildQueueWebsocketService.sendBuildAgentInformation(getBuildAgentInformation());
+            }
         }
 
         @Override
         public void entryRemoved(com.hazelcast.core.EntryEvent<String, LocalCIBuildAgentInformation> event) {
-            log.debug("Build agent removed: {}", event.getOldValue());
-            localCIBuildQueueWebsocketService.sendBuildAgentInformation(getBuildAgentInformation());
+            if (schedulingActive) {
```
Comment (mateusmm01) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCISharedBuildJobQueueService.java:483]: Perhaps create an extra method sendBuildAgentInformationOverWebsocket() as you did in the other Listeners for consistency

**Review by reschandreas — CHANGES_REQUESTED** [review]
some nitpicking comments, but please also update `artemis.continuous-integration.thread-pool-size=1` in `AbstractSpringIntegrationLocalCILocalVCTest` line 68

**Review by mateusmm01 — DISMISSED** [review]
Code LGTM. Left a small comment

**Review by reschandreas — APPROVED** [review]
approve after code changes

**Review by mateusmm01 — APPROVED** [review]
Code changes look good !

**Review by milljoniaer — APPROVED** [review]
code looks good 🚀

**Review by egekurt123 — APPROVED** [review]
Tested manually on ts6 works as expected for Java exercise, but when I submitted for any C exercise, it didn't show up in the build queue. I don't think this problem is because of the changes in this PR.

**Comment by laurenzfb** [comment]
@egekurt123 C exercisesbuild really fast, they might be too quick to show up
